File: leetcode/medium/DeleteNodeInABST.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        """
        Deletes a node with the given key from a Binary Search Tree (BST).

        The function recursively traverses the BST to find the node to delete.
        Once the node is found, it handles the deletion based on three cases:
        1. The node is a leaf (no children): It is simply removed.
        2. The node has one child: The node is replaced by its child.
        3. The node has two children: The node's value is replaced by the value of its
           in-order successor (the smallest value in its right subtree). Then, the
           in-order successor is recursively deleted from the right subtree.

        Args:
            root: The root of the BST.
            key: The key of the node to be deleted.

        Returns:
            The root of the modified BST.
        """
        # Base case: If the tree is empty, return None.
        if not root:
            return None

        # Recursively search for the node to delete.
        if key < root.val:
            # The key is in the left subtree.
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            # The key is in the right subtree.
            root.right = self.deleteNode(root.right, key)
        else:
            # The current node is the one to be deleted.

            # Case 1: Node has no left child.
            # Replace the node with its right child (or None if it has no right child).
            if not root.left:
                return root.right
            
            # Case 2: Node has no right child.
            # Replace the node with its left child.
            elif not root.right:
                return root.left
            
            # Case 3: Node has two children.
            # Find the in-order successor (the smallest node in the right subtree).
            successor = self._find_min(root.right)
            
            # Replace the current node's value with the successor's value.
            root.val = successor.val
            
            # Recursively delete the successor from the right subtree.
            root.right = self.deleteNode(root.right, successor.val)
            
        return root

    def _find_min(self, node: TreeNode) -> TreeNode:
        """
        Finds the node with the minimum value in a given subtree (the leftmost node).
        """
        current = node
        while current and current.left:
            current = current.left
        return current

File: leetcode/medium/DeleteNodeInABST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    /**
     * Deletes a node with the given key from the Binary Search Tree (BST).
     * The method handles three cases:
     * 1. Node is not found (root is null).
     * 2. Node has 0 or 1 child.
     * 3. Node has 2 children (replaces node's value with its inorder successor).
     *
     * @param root The root of the current (sub)tree.
     * @param key The value of the node to be deleted.
     * @return The root of the (sub)tree after deletion.
     */
    public TreeNode deleteNode(TreeNode root, int key) {
        // Base case: If the tree is empty or key is not found
        if (root == null) {
            return null;
        }

        // 1. Search for the node to delete (standard BST search)
        if (key < root.val) {
            // Traverse left subtree
            root.left = deleteNode(root.left, key);
        } else if (key > root.val) {
            // Traverse right subtree
            root.right = deleteNode(root.right, key);
        }
        // 2. Found the node to be deleted (key == root.val)
        else {
            // Case 1: Node has 0 or 1 child
            
            // If the node has no left child (or is a leaf node)
            if (root.left == null) {
                // Replace the current node with its right child (could be null)
                return root.right; 
            }
            // If the node has no right child
            if (root.right == null) {
                // Replace the current node with its left child
                return root.left;
            }

            // Case 2: Node has 2 children
            
            // Find the inorder successor (smallest node in the right subtree)
            int inorderSuccessorVal = findInorderSuccessor(root);
            
            // Replace the current node's value with the successor's value
            root.val = inorderSuccessorVal;
            
            // Delete the original inorder successor from the right subtree
            // The successor is now redundant as its value has been moved up.
            root.right = deleteNode(root.right, inorderSuccessorVal);
        }
        
        // Return the (potentially modified) root of the current subtree
        return root;
    }

    /**
     * Finds the value of the inorder successor for a node that has two children.
     * The inorder successor is the smallest value in the node's right subtree.
     *
     * @param rootToDelete The node whose successor needs to be found. 
     * It must have a non-null right child.
     * @return The integer value of the inorder successor.
     */
    private int findInorderSuccessor(TreeNode rootToDelete) {
        // Start from the right child of the node to be deleted
        TreeNode tempNode = rootToDelete.right;
        
        // The inorder successor is the leftmost node in the right subtree
        while (tempNode.left != null) {
            tempNode = tempNode.left;
        }
        return tempNode.val;
    }
}